File: packages/supreme-sdk-dart/test/client_test.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:supreme_sdk/supreme_sdk.dart';
import 'package:test/test.dart';

/// The access token is short-lived (15 min) by design; every command must survive it expiring
/// mid-session by silently refreshing and retrying — never leaving a device stuck (or, worse, the
/// app looking "logged in" while every request quietly 401s). This is the fix for a bug where the
/// Dart SDK never refreshed at all.
http.Response _json(int status, Map<String, dynamic> body) =>
    http.Response(jsonEncode(body), status, headers: {'content-type': 'application/json'});

void main() {
  test('refreshes once and retries a command that 401s on an expired access token', () async {
    final paths = <String>[];
    final client = SupremeClient(
      baseUrl: 'http://hub.local',
      httpClient: MockClient((req) async {
        paths.add(req.url.path);
        final auth = req.headers['authorization'];
        if (req.url.path == '/v1/devices/dev1/command' && auth == 'Bearer stale') {
          return _json(401, {'code': 'unauthorized', 'message': 'token expired'});
        }
        if (req.url.path == '/v1/auth/refresh') {
          return _json(200, {'accessToken': 'fresh', 'refreshToken': 'refresh2', 'expiresIn': 900, 'tokenType': 'Bearer'});
        }
        if (req.url.path == '/v1/devices/dev1/command' && auth == 'Bearer fresh') {
          return _json(200, {'accepted': true});
        }
        throw StateError('unexpected request: ${req.url.path}');
      }),
    );
    client.restoreTokens(accessToken: 'stale', refreshToken: 'refresh1');

    await client.command('dev1', {'capability': 'onoff', 'action': 'on'});

    expect(paths, ['/v1/devices/dev1/command', '/v1/auth/refresh', '/v1/devices/dev1/command']);
    expect(client.accessToken, 'fresh');
  });

  test('shares one refresh across concurrent 401s instead of racing the refresh endpoint', () async {
    var refreshCalls = 0;
    final client = SupremeClient(
      baseUrl: 'http://hub.local',
      httpClient: MockClient((req) async {
        if (req.url.path == '/v1/auth/refresh') {
          refreshCalls += 1;
          return _json(200, {'accessToken': 'fresh', 'refreshToken': 'refresh2', 'expiresIn': 900, 'tokenType': 'Bearer'});
        }
        if (req.headers['authorization'] == 'Bearer stale') return _json(401, {'code': 'unauthorized', 'message': 'expired'});
        return _json(200, {'accepted': true});
      }),
    );
    client.restoreTokens(accessToken: 'stale', refreshToken: 'refresh1');

    await Future.wait([
      client.command('dev1', {'capability': 'onoff', 'action': 'on'}),
      client.command('dev2', {'capability': 'onoff', 'action': 'off'}),
      client.command('dev3', {'capability': 'onoff', 'action': 'toggle'}),
    ]);

    expect(refreshCalls, 1);
  });

  test('persists every rotated refresh token, not just the one from login', () async {
    final persisted = <String>[];
    final client = SupremeClient(
      baseUrl: 'http://hub.local',
      onTokensChanged: (access, refresh) => persisted.add(refresh),
      httpClient: MockClient((req) async {
        if (req.url.path == '/v1/auth/refresh') {
          return _json(200, {'accessToken': 'fresh', 'refreshToken': 'rotated-2', 'expiresIn': 900, 'tokenType': 'Bearer'});
        }
        if (req.headers['authorization'] == 'Bearer stale') return _json(401, {'code': 'unauthorized', 'message': 'expired'});
        return _json(200, {'accepted': true});
      }),
    );
    client.restoreTokens(accessToken: 'stale', refreshToken: 'rotated-1');

    await client.command('dev1', {'capability': 'onoff', 'action': 'on'});

    // The ROTATED token must be captured — persisting only the login-time token would get the
    // session revoked (reuse-detected) the next time the app reads its stale persisted value.
    expect(persisted, ['rotated-2']);
    expect(client.refreshToken, 'rotated-2');
  });

  test('clears the session and reports it as expired when the refresh token is also dead', () async {
    var expired = false;
    final client = SupremeClient(
      baseUrl: 'http://hub.local',
      onSessionExpired: () => expired = true,
      httpClient: MockClient((req) async {
        if (req.url.path == '/v1/auth/refresh') return _json(401, {'code': 'unauthorized', 'message': 'refresh token revoked'});
        return _json(401, {'code': 'unauthorized', 'message': 'token expired'});
      }),
    );
    client.restoreTokens(accessToken: 'stale', refreshToken: 'dead');

    await expectLater(client.command('dev1', {'capability': 'onoff', 'action': 'on'}), throwsA(isA<SupremeApiException>()));
    expect(expired, isTrue);
    expect(client.accessToken, isNull);
  });

  test('an explicit refresh() call also fires onSessionExpired on failure (not just the 401 retry path)', () async {
    var expired = false;
    final client = SupremeClient(
      baseUrl: 'http://hub.local',
      onSessionExpired: () => expired = true,
      httpClient: MockClient((req) async => _json(401, {'code': 'unauthorized', 'message': 'refresh token revoked'})),
    );
    client.restoreTokens(accessToken: 'stale', refreshToken: 'dead');

    await expectLater(client.refresh(), throwsA(isA<SupremeApiException>()));
    expect(expired, isTrue);
  });
}

File: packages/supreme-sdk-ts/src/client.test.ts
import type { DeviceId } from "@supreme/domain-model";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { MemoryTokenStore, SupremeClient } from "./client.js";

/** Builds an unsigned-but-shaped JWT carrying only the `exp` claim the SDK actually reads. */
function fakeJwt(expSeconds: number): string {
  const b64 = (obj: unknown) => btoa(JSON.stringify(obj)).replace(/\+/g, "-").replace(/\//g, "_");
  return `${b64({ alg: "none" })}.${b64({ exp: expSeconds })}.sig`;
}

/**
 * The access token is short-lived (15 min) by design; every homeowner command must survive it
 * expiring mid-session by silently refreshing and retrying — never surfacing a stale-session error
 * for something that's just a routine token rotation. This is the fix for a bug where the SDK never
 * refreshed at all, so every request started failing 401 (silently, since call sites `void` the
 * promise) once 15 minutes had passed since login — indistinguishable from "the app stopped working".
 */
function jsonResponse(status: number, body: unknown): Response {
  return new Response(JSON.stringify(body), { status, headers: { "content-type": "application/json" } });
}

describe("SupremeClient — automatic refresh-and-retry on 401", () => {
  it("refreshes once and retries a command that 401s on an expired access token", async () => {
    const calls: { path: string; auth: string | null }[] = [];
    const fetchImpl = (async (url: string, init?: RequestInit) => {
      const path = new URL(url).pathname;
      const auth = (init?.headers as Record<string, string>)?.authorization ?? null;
      calls.push({ path, auth });
      if (path === "/v1/devices/dev1/command" && auth === "Bearer stale") {
        return jsonResponse(401, { code: "unauthorized", message: "token expired" });
      }
      if (path === "/v1/auth/refresh") {
        return jsonResponse(200, { accessToken: "fresh", refreshToken: "refresh2", expiresIn: 900, tokenType: "Bearer" });
      }
      if (path === "/v1/devices/dev1/command" && auth === "Bearer fresh") {
        return jsonResponse(200, { accepted: true });
      }
      throw new Error(`unexpected request: ${path}`);
    }) as typeof fetch;

    const tokenStore = new MemoryTokenStore();
    tokenStore.set({ accessToken: "stale", refreshToken: "refresh1" });
    const client = new SupremeClient({ baseUrl: "http://hub.local", tokenStore, fetchImpl });

    const result = await client.command("dev1" as DeviceId, { capability: "onoff", action: "on" });

    expect(result).toEqual({ accepted: true });
    expect(calls.map((c) => c.path)).toEqual([
      "/v1/devices/dev1/command", // first attempt, 401s
      "/v1/auth/refresh", // silent refresh
      "/v1/devices/dev1/command", // retried with the fresh token
    ]);
    expect(client.accessToken).toBe("fresh");
  });

  it("shares one refresh across concurrent 401s instead of racing the refresh endpoint", async () => {
    let refreshCalls = 0;
    const fetchImpl = (async (url: string, init?: RequestInit) => {
      const path = new URL(url).pathname;
      const auth = (init?.headers as Record<string, string>)?.authorization ?? null;
      if (path === "/v1/auth/refresh") {
        refreshCalls += 1;
        return jsonResponse(200, { accessToken: "fresh", refreshToken: "refresh2", expiresIn: 900, tokenType: "Bearer" });
      }
      if (auth === "Bearer stale") return jsonResponse(401, { code: "unauthorized", message: "expired" });
      return jsonResponse(200, { accepted: true });
    }) as typeof fetch;

    const tokenStore = new MemoryTokenStore();
    tokenStore.set({ accessToken: "stale", refreshToken: "refresh1" });
    const client = new SupremeClient({ baseUrl: "http://hub.local", tokenStore, fetchImpl });

    await Promise.all([
      client.command("dev1" as DeviceId, { capability: "onoff", action: "on" }),
      client.command("dev2" as DeviceId, { capability: "onoff", action: "off" }),
      client.command("dev3" as DeviceId, { capability: "onoff", action: "toggle" }),
    ]);

    expect(refreshCalls).toBe(1);
  });

  it("clears the session and reports it as expired when the refresh token is also dead", async () => {
    let expired = false;
    const fetchImpl = (async (url: string) => {
      const path = new URL(url).pathname;
      if (path === "/v1/auth/refresh") return jsonResponse(401, { code: "unauthorized", message: "refresh token revoked" });
      return jsonResponse(401, { code: "unauthorized", message: "token expired" });
    }) as typeof fetch;

    const tokenStore = new MemoryTokenStore();
    tokenStore.set({ accessToken: "stale", refreshToken: "dead" });
    const client = new SupremeClient({
      baseUrl: "http://hub.local",
      tokenStore,
      fetchImpl,
      onSessionExpired: () => { expired = true; },
    });

    await expect(client.command("dev1" as DeviceId, { capability: "onoff", action: "on" })).rejects.toThrow();
    expect(expired).toBe(true);
    expect(client.accessToken).toBeNull();
  });
});

/**
 * BUG-003: reactive (on-401) refresh alone means every request in flight the instant the access
 * token expires 401s together — a dashboard mount fires a dozen-plus GETs, so that was the "15-18
 * failed requests before the session recovers" symptom. The SDK now also refreshes proactively,
 * ~60s ahead of the token's own `exp` claim, so a session that's actively in use never reaches
 * that burst in the first place.
 */
describe("SupremeClient — proactive refresh ahead of token expiry", () => {
  beforeEach(() => vi.useFakeTimers());
  afterEach(() => vi.useRealTimers());

  it("refreshes on its own, before any request ever sees a 401", async () => {
    const nowSec = Math.floor(Date.now() / 1000);
    let refreshCalls = 0;
    let sawAny401 = false;
    const fetchImpl = (async (url: string) => {
      const path = new URL(url).pathname;
      if (path === "/v1/auth/refresh") {
        refreshCalls += 1;
        return jsonResponse(200, {
          accessToken: fakeJwt(nowSec + 900),
          refreshToken: "refresh2",
          expiresIn: 900,
          tokenType: "Bearer",
        });
      }
      sawAny401 = true; // any other call in this test is unexpected
      return jsonResponse(401, { code: "unauthorized", message: "unexpected call" });
    }) as typeof fetch;

    const tokenStore = new MemoryTokenStore();
    tokenStore.set({ accessToken: fakeJwt(nowSec + 90), refreshToken: "refresh1" });
    new SupremeClient({ baseUrl: "http://hub.local", tokenStore, fetchImpl });

    // Token expires in 90s and the buffer is 60s, so the proactive timer should fire at ~30s.
    await vi.advanceTimersByTimeAsync(31_000);

    expect(refreshCalls).toBe(1);
    expect(sawAny401).toBe(false);
  });

  it("refreshes immediately on construction when a restored session's token is already past the buffer", async () => {
    const nowSec = Math.floor(Date.now() / 1000);
    let refreshCalls = 0;
    const fetchImpl = (async (url: string) => {
      const path = new URL(url).pathname;
      if (path === "/v1/auth/refresh") {
        refreshCalls += 1;
        return jsonResponse(200, {
          accessToken: fakeJwt(nowSec + 900),
          refreshToken: "refresh2",
          expiresIn: 900,
          tokenType: "Bearer",
        });
      }
      return jsonResponse(401, { code: "unauthorized", message: "unexpected call" });
    }) as typeof fetch;

    const tokenStore = new MemoryTokenStore();
    tokenStore.set({ accessToken: fakeJwt(nowSec + 10), refreshToken: "refresh1" }); // inside the 60s buffer already
    new SupremeClient({ baseUrl: "http://hub.local", tokenStore, fetchImpl });

    await vi.advanceTimersByTimeAsync(0);

    expect(refreshCalls).toBe(1);
  });
});
